cli: fix dump to gracefully handle temp tables, views and sequences

adityamaru opened the pull request:
pg_dump ignores temporary tables, views and sequences.  This change
makes `cockroach dump` follow the same behavior by skipping descriptors
with schema name having `pg_temp` as its prefix, when aggregating the
metadata for the tables to be dumped.

When no table names are specified in the `cockroach dump` command, we
silently ignore all temp objects.  If the user explicitly asks for a
temp object to be dumped, we error out with an informative message.

Fixes:#50899
Fixes:#51166

Release note (bug fix): `cockroach dump` no longer errors out when
dumping temporary tables, view or sequences.  It either ignores them or
throws an informative error if the temp object is explicitly requested
to be dumped via the CLI.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51185)
<!-- Reviewable:end -->

rohany commented on pkg/cli/dump.go:
```diff
@@ -532,26 +536,41 @@ func getBasicMetadata(conn *sqlConn, dbName, tableName string, ts string) (basic
+	if !ok {
+		return basicMetadata{}, fmt.Errorf("unexpected value: %T", schemaNameI)
+	}
+	if strings.Contains(schemaName, sessiondata.PgTempSchemaName) {
```
not contains, but `Prefix`

rohany commented on pkg/cli/dump_test.go:
```diff
@@ -858,3 +860,171 @@ INSERT INTO foo (id) VALUES
 	}
 }
+
+func TestDumpTempTables(t *testing.T) {
```
use the datadriven dump tests instead -- take a look at `pkg/cli/testdata/dump/`

rohany replied:
I would just comment here why you didn't use the datadriven tests so that we remember

rohany commented:
This PR should be backported to 20.1

donbowman commented:
I guess if this is accepted #51173 becomes moot.

donbowman commented:
One thing that nags at me
https://github.com/cockroachdb/cockroach/issues/42425
suggests that the temp tables/sequences/... are deleted when no longer in use (session gone).
but that does not seem to be the case:
```
select distinct schema_name from crdb_internal.create_statements;
```
shows 25 items
show sessions reports 1 (the session i have open doing the show sessions)

Are they not supposed to be removed from here?

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51185)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @rohany)

___
*[pkg/cli/dump.go, line 550 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51185#-MBlM94kpAMBYx4HMRUI-r1-550:-MBnfVXqCN73BgXcN1JN:b1jxzi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/51974f79a23713a673d7399739295eef52df50d7/pkg/cli/dump.go#L550)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

not contains, but `Prefix`
</blockquote></details>

good catch, done.
___
*[pkg/cli/dump_test.go, line 864 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51185#-MBlM94kpAMBYx4HMRUJ-r1-864:-MBnxV6G9jJ0QPDFhq7P:b-brv130) ([raw file](https://github.com/cockroachdb/cockroach/blob/51974f79a23713a673d7399739295eef52df50d7/pkg/cli/dump_test.go#L864)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

use the datadriven dump tests instead -- take a look at `pkg/cli/testdata/dump/`
</blockquote></details>

as discussed offline, temp tables do not persist across `RunWithCaptureArgs()` invocations. Therefore, using the datadriven infrastructure does not allow us to test our handling of temp object since they get deleted between `sql` and `dump` runs anyways.


<!-- Sent from Reviewable.io -->

rohany commented:
> suggests that the temp tables/sequences/... are deleted when no longer in use (session gone).

They are cleaned up periodically by a background process, not immediately as the session exits.

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51185)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @rohany)

___
*[pkg/cli/dump_test.go, line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51185#-MBlM94kpAMBYx4HMRUJ-r2-864:-MBo05KO0Kkjg4Lca1gO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da968963457923b358fa7f8b7dd98075845a5933/pkg/cli/dump_test.go#L864)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I would just comment here why you didn't use the datadriven tests so that we remember
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adityamaru commented:
Can't reproduce the CI stress fail locally or on gceworker. I suspect it could be because the PGURL name was a constant rather than the convention which is `t.Name`.

adityamaru commented:
Pushed to get a CI run in. Testing whether the time statement in main_test.go is the cause of stressrace failure on CI. The test is now passing on gceworker.

adityamaru reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51185)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @rohany)

___
*[pkg/cli/main_test.go, line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51185#-MBqIiIz6PTXOnbbve5E:-MBqIiIz6PTXOnbbve5F:bysppdq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bc2b26eab8fc9affcbfbda513c9ba6c1ffa51d0/pkg/cli/main_test.go#L36)):*
> ```
> 
> ```
@raduberinde keeping this line is causing the newly added `TestDumpTempTables` to fail under `make stress` with - https://teamcity.cockroachdb.com/repository/download/Cockroach_UnitTests/2078378:id/testrace/logTestDumpTempTables_dump_db_only_temp068989627/clitest-stderr.4ab17fe2fbed.roach.2020-07-09T20_12_24Z.018111.log

It also causes the other older unit tests to fail when I run them on a gceworker, with the same error as linked above. I tried moving the time statement into `init` based on a thread I saw on Slack, but that doesn't work either. Any recommendations on how to remedy this? 


<!-- Sent from Reviewable.io -->

knz commented on pkg/cli/main_test.go:
```diff
@@ -31,10 +30,6 @@ func TestMain(m *testing.M) {
 	// a version injected. Pretend to be a very up-to-date version.
 	defer build.TestingOverrideTag("v999.0.0")()
 
-	// The times for Example_statement_diag are reported in the local timezone.
```
Why did you change this?

adityamaru replied:
I didn't plan to check-in the deletion, I just wanted to see if that was the cause of the`make stress` failures for the dump unit tests on CI. I have asked Radu for his input on the same in the comment above - https://github.com/cockroachdb/cockroach/pull/51185#pullrequestreview-446066146

adityamaru committed:
cli: change statement diag to use UTC time instead of Local time

Previously, in the case of cli tests we overrode Local to be equal to
UTC time so that the output is always the same. This assignment however
resulted in data race conditions in several CLI unit tests.

Egs:
https://teamcity.cockroachdb.com/repository/download/Cockroach_UnitTests/
2078378:id/testrace/logTestDumpTempTables_dump_db_only_temp068989627/
clitest-stderr.4ab17fe2fbed.roach.2020-07-09T20_12_24Z.018111.log

This change ensures all statement diag times are always show in UTC, thereby
eliminating the need to override Local.

Release note (cli change): statement-diag command will now show all
times in UTC.

adityamaru committed:
cli: fix dump to gracefully handle temp tables, views and sequences

pg_dump ignores temporary tables, views and sequences.  This change
makes `cockroach dump` follow the same behavior by skipping descriptors
with schema name having `pg_temp` as its prefix, when aggregating the
metadata for the tables to be dumped.

When no table names are specified in the `cockroach dump` command, we
silently ignore all temp objects.  If the user explicitly asks for a
temp object to be dumped, we error out with an informative message.

Fixes:#50899

Release note (bug fix): `cockroach dump` no longer errors out when
dumping temporary tables, view or sequences.  It either ignores them or
throws an informative error if the temp object is explicitly requested
to be dumped via the CLI.

adityamaru commented:
@RaduBerinde as discussed offline I changed Local() to UTC() in the statement_diag. I have separated the change out into its own commit.

RaduBerinde commented:
Statement diag commit LGTM, thanks!

rohany commented:
@adityamaru do you know why this was segfaulting though with the `Local` though?

RaduBerinde commented:
Looks like it was racing with Go's timer code which sends `Now()` through some channel. It's enough for any other package to have set up a timer in an `init()`. Modifying `Local` is a pretty broken "API".

adityamaru commented:
TFTRs!

bors r+

otan commented:
should we backport this to v20.1?

knz commented:
it's a bug fix without backward-incompatibility so yes
